src/unix.rs

```diff
@@ -17,17 +17,15 @@ use std::thread::{self, Builder, JoinHandle};
 use std::time::Duration;
 
 #[derive(Debug)]
-pub enum Client {
-    /// `--jobserver-auth=R,W`
-    Pipe { read: File, write: File },
-    /// `--jobserver-auth=fifo:PATH`
-    Fifo {
-        file: File,
-        path: PathBuf,
-        /// it can only go from false -> true but not the other way around, since that
-        ///  could cause a race condition.
-        is_non_blocking: AtomicBool,
-    },
+pub struct Client {
+    read: File,
+    write: Option<File>,
+    /// If path is not None, then a fifo jobserver is created.
+    path: Option<Box<Path>>,
+    supports_non_blocking: bool,
+    /// it can only go from false -> true but not the other way around, since that
+    ///  could cause a race condition.
+    is_non_blocking: AtomicBool,
```

Comment:
The new structure is confusing. `read` sometimes also means write.
The old structure was maybe a little bit more readable, covering different cases explicitly, but it was still a lossy transform of what the parent passed.

Additionally it does not track well whether we have inherited things from a parent or crated our own, which can affect how we should handle environment variables (which could be relevant to #64)

We should do more than fix a bug, we should also keep things organized.

Reply:
> Additionally it does not track well whether we have inherited things from a parent or crated our own

This js tracked by the `path` variable, if it is `Some` then we inherited/created a fifo, or we inherited/created a pipe.

Reply:
> The new structure is confusing. read sometimes also means write.

I could convert that back to an `enum`

Reply:
The thing is we're tracking a bunch of flags and implied state that affects other fields. That obfuscates state.
As I said, the old enum was a bit better, but not good enough either since it did not to cover some state.

I think there are two different aspects
* what we can use ourselves
  * which might be 2 fds, 1 fifo or 2fds+1fifo
  * the blockingness
* what we give to children

Reply:
I think maybe we should unify the fifo/pipe?

```rust
struct Client {
    read: File,
    write: File,
    fifo_path: Option<Box<Path>>,
    is_blocking: Option<AtomicBool>,
}
```

It would require an additional open for fifo and an additional fd, but guaranteed it reduces complexity maybe we should adopt it?

It would also separate the read/write, and make sure setting non-blocking on read end would not affect the write end.

Reply:
> I think maybe we should unify the fifo/pipe?

I don't think we should do that. In the case where we inherit 2 FDs but reopen it via procfs it would make the unused FD numbers (which we would still pass to children) float around implicitly.
We should encode this explicitly. `std` now has a debug assert that will try to check that owned FDs are not closed twice. So if `jobserver` claims ownership of an FD this will be checked in debug mode.

Reply:
No I mean we should open the fifo twice:

```rust
let read = open_read(fifo_path?);
let write = open_write(fifo_path?);

Client {
    read,
    write,
    ...
}
```

Then when we set `O_NONBLOCK` on the read end, the write would be still in blocking mode instead of being changed to non-blocking mode.

Reply:
That'd could work for the "real" fifo case but not for the optimization where we inherit 2 fds (and keep passing those to children) and additionally open a process-private one.

Reply:
> but not for the optimization where we inherit 2 fds

I think it should work, previously we use `/dev/fd/$fd` to reopen the read fd as readable and writable.

Better solution would be to re-open read end as readable, write end as writable.

I would try this locally and see if it works.

Reply:
My point is that we should _also_ keep track of the original FDs because those are the ones we're going to pass to the children.

Reply:
Aha right, yes we should do that, because we set the non-blocking on read end.

Reply:
So, I propose something like:

```rust
enum ClientSource {
    Fds {
        read: RawFd,
        write: RawFd,
    },
    Fifo(Box<Path>),
}

struct Client {
    read: File,
    write: File,
    is_non_blocking: Option<AtomicBool>,
    source: ClientSource,
}
```

Reply:
@the8472 Does that look good to you?

Reply:
Hrrm, it looks better. We probably should keep track of the environment too but that's not related to the issue this PR is trying to fix so it can happen elsewhere.

I'll have to see the read/write ends in action to see if it makes sense.

Reply:
Updated, turns out for the real fifo, I can open it only once within a process with RW, if I open it with readonly, then open it with write only would block, so I instead clone the fd.

For the optimization of creating a fifo from pipe, it seems to work well.

Reply:
> Updated, turns out for the real fifo, I can open it only once within a process with RW

Wouldn't that be a problem for the real fifo case? `Client::from_env` docs say

> It is ok to call this function any number of times.

If this would block that could cause issues for the caller.

Though I'm not seeing the "open twice" scenario mentioned in `man 7 fifo`. Maybe `open(..., O_NONBLOCK)` and then unsetting it afterwards via `fcntl` would help?

Reply:
> Wouldn't that be a problem for the real fifo case? Client::from_env docs say

In the test, the `Client` is the only one opening the fifo.

https://stackoverflow.com/questions/76411975/is-there-a-way-to-open-fifonamed-pipe-in-write-only-way-without-blocked

> From [the FIFO manual page](https://man7.org/linux/man-pages/man7/fifo.7.html):
>
> > Normally, opening the FIFO blocks until the other end is opened also.

So I guess open it with read + write makes sense and prevent that blocking.

Reply:
> Maybe open(..., O_NONBLOCK) and then unsetting it afterwards via fcntl would help?

Yeah that should help:

> A process can open a FIFO in nonblocking mode.  In this case,
       opening for read-only succeeds even if no one has opened on the
       write side yet and opening for write-only fails with ENXIO (no
       such device or address) unless the other end has already been
       opened.

Reply:
> We can open the fifo again with read/write only.

Oh ok, then I misunderstood what you wrote.

In that case we could open it twice instead of cloning it, that way we could even manage the O_NONBLOCK flags separately.

Reply:
Thanks, changed it to open it twice.